hail/hail/ir-gen/src/Main.scala

```diff
@@ -0,0 +1,1239 @@
+    x1: ConstrainedType[T1],
+    x2: ConstrainedType[T2],
+    x3: ConstrainedType[T3],
+  ) extends Tup[(T1, T2, T3)](x1, x2, x3)
+
+  case class NodeDef(
+    name: String,
+    attsAndChildren: Seq[Declaration[_]],
+    traits: Seq[Trait] = Seq.empty,
+    constraints: Seq[String] = Seq.empty,
+    extraMethods: Seq[String] = Seq.empty,
+    staticMethods: Seq[String] = Seq.empty,
+    docstring: String = "",
+    hasCompanionExtension: Boolean = false,
+  ) extends NodeDefInterface {
+    override def withTraits(newTraits: Trait*): NodeDef = copy(traits = traits ++ newTraits)
+
+    override def withMethod(methodDef: String): NodeDef =
+      copy(extraMethods = extraMethods :+ methodDef)
+
+    override def typed(typ: String): NodeDef = withTraits(TypedIR(typ))
+    override def withConstraint(c: String): NodeDef = copy(constraints = constraints :+ c)
+
+    override def withCompanionExtension: NodeDef = copy(hasCompanionExtension = true)
+
+    override def withClassExtension: NodeDef = withTraits(Trait(s"${name}Ext"))
+
+    override def withDocstring(docstring: String): NodeDef = copy(docstring = docstring)
+
+    private def nChildren: IntRepr = attsAndChildren.foldLeft(IntRepr())(_ + _.nChildren)
+
+    private def children: String =
+      attsAndChildren.foldLeft[SeqRepr[Child]](SeqRepr.empty(Child_("BaseIR")))(
+        _ ++ _.childrenSeq
+      ).toString
+
+    private def childrenOffsets: Seq[IntRepr] =
+      attsAndChildren.scanLeft(IntRepr())(_ + _.nChildren)
+
+    private def copyMethod: String = {
+      val decl = s"override def copyWithNewChildren(newChildren: IndexedSeq[BaseIR]): $name = "
+      val assertion = s"assert(newChildren.length == $nChildren)"
+      val body = name + (attsAndChildren, childrenOffsets).zipped.map { case (x, offset) =>
+        val newChildren = SeqRepr.Dynamic("newChildren", Child_("BaseIR"))
+        x.repr.copyWithNewChildren(
+          if (x.nChildren.hasStaticValue(0)) SeqRepr.empty(Child_("BaseIR"))
+          else if (x.nChildren.hasStaticValue(1))
+            SeqRepr.Static(Seq(newChildren(offset)), Child_("BaseIR"))
+          else if (x.nChildren == nChildren) newChildren
+          else newChildren.slice(offset, x.nChildren)
+        )
+      }.mkString("(", ", ", ")")
+      decl + "{\n    " + assertion + "\n    " + body + "}"
+    }
+
+    private def classDecl = {
+      val paramList = s"$name(${attsAndChildren.map(_.generateDeclaration).mkString(", ")})"
+      s"final case class $paramList extends IR" + traits.map(" with " + _.name).mkString
+    }
+
+    private def classBody = {
+      val childrenSeqDef = if (nChildren.hasStaticValue(0))
+        s"override def childrenSeq: IndexedSeq[BaseIR] = IndexedSeq.empty"
+      else
+        s"override lazy val childrenSeq: IndexedSeq[BaseIR] = $children"
+      val extraMethods =
+        this.extraMethods :+ childrenSeqDef :+ copyMethod
+      val constraints = this.constraints ++ attsAndChildren.flatMap(_.constraints)
```

Comment:
Not a huge issue, I think of "constraints" as being the type constraints in class or method definitions like
```scala
def launchTheAribitraryMissiles[A: Arbitrary]: Unit
```

Since you're using these to generate "require" statements, perhaps call them "requirements"? Or just "valueConstraints"?